### ISSUE TITLE
ExaBGP crashes then IPv6 session is taken down from the remote side

### ISSUE THREAD

**Issue by pettai** [Exa-Networks/exabgp#912]
##### Summary
ExaBGP crashes then IPv6 BGP session is taken down (cleared) from the remote side

##### OS
```
Linux Ubuntu 18.04
4.15.0-52-generic #56-Ubuntu SMP Tue Jun 4 22:49:08 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

##### Version
Latest version from Debian/Ubuntu's apt repo
```
ExaBGP : 4.0.8-793a2931
Python : 3.6.8 (default, Jan 14 2019, 11:02:34)  [GCC 8.0.1 20180414 (experimental) [trunk revision 259383]]
```

##### Installation
```
latest OS package
```

##### Environment 
```
[exabgp.daemon]
daemonize = true
user = 'exabgp'
```

##### Configuration
```
neighbor 130.242.X.X {
	description "r3-router";
	router-id <router-id>;
	local-address 130.242.Y.Y;
	local-as <localASN>;
	peer-as 1653;
}

neighbor 2001:6b0::xx {
	description "r3-router";
	router-id <router-id>;
	local-address 2001:6b0::yy;
	local-as <localASN>;
	peer-as 1653;
}

process watch-kresd {
	run python3 -m exabgp healthcheck --config /etc/exabgp/healthcheck-kresd.conf;
	encoder text;
}
```

##### Program output
```
Aug 23 17:28:42 punoqun exabgp[16044]: ********************************************************************************
exabgp[16044]: -- Please provide _ALL_ the information above on :
exabgp[16044]: -- https://github.com/Exa-Networks/exabgp/issues
exabgp[16044]: ********************************************************************************
exabgp[16044]: Traceback (most recent call last):
exabgp[16044]:   File "/usr/lib/python3/dist-packages/exabgp/reactor/peer.py", line 535, in _run
exabgp[16044]:     for action in self._main():
exabgp[16044]:   File "/usr/lib/python3/dist-packages/exabgp/reactor/peer.py", line 406, in _main
exabgp[16044]:     for message in self.proto.read_message():
exabgp[16044]:   File "/usr/lib/python3/dist-packages/exabgp/reactor/protocol.py", line 249, in read_message
exabgp[16044]:     raise message
exabgp[16044]: exabgp.bgp.message.notification.Notification: <unprintable Notification object>
exabgp[16044]: During handling of the above exception, another exception occurred:
exabgp[16044]: Traceback (most recent call last):
exabgp[16044]:   File "/usr/sbin/exabgp", line 11, in <module>
exabgp[16044]:     load_entry_point('exabgp==4.0.8', 'console_scripts', 'exabgp')()
exabgp[16044]:   File "/usr/lib/python3/dist-packages/exabgp/application/__init__.py", line 13, in run_exabgp
exabgp[16044]:     main()
exabgp[16044]:   File "/usr/lib/python3/dist-packages/exabgp/application/bgp.py", line 302, in main
exabgp[16044]:     run(env,comment,configurations,root,options["--validate"])
exabgp[16044]:   File "/usr/lib/python3/dist-packages/exabgp/application/bgp.py", line 368, in run
exabgp[16044]:     exit_code = Reactor(configurations).run(validate,root)
exabgp[16044]:   File "/usr/lib/python3/dist-packages/exabgp/reactor/loop.py", line 259, in run
exabgp[16044]:     action = peer.run()
exabgp[16044]:   File "/usr/lib/python3/dist-packages/exabgp/reactor/peer.py", line 626, in run
exabgp[16044]:     return six.next(self.generator)
exabgp[16044]:   File "/usr/lib/python3/dist-packages/exabgp/reactor/peer.py", line 577, in _run
exabgp[16044]:     notification
exabgp[16044]:   File "/usr/lib/python3/dist-packages/exabgp/reactor/peer.py", line 133, in _reset
exabgp[16044]:     self.proto.close(u"peer reset, message [{0}] error[{1}]".format(message, error))
exabgp[16044]:   File "/usr/lib/python3/dist-packages/exabgp/bgp/message/notification.py", line 155, in __str__
exabgp[16044]:     '%s' % (' / %s' % str_ascii(self.data) if self.data else '')
exabgp[16044]:   File "/usr/lib/python3/dist-packages/exabgp/util/__init__.py", line 84, in str_ascii
exabgp[16044]:     return str(bytestring, 'ascii')
exabgp[16044]: UnicodeDecodeError: 'ascii' codec can't decode byte 0xb0 in position 6: ordinal not in range(128)
```

##### Steps to reproduce
* setup a IPv4+IPv6 bgp session between exabgp <-> (juniper) router
* clear the IPv6 session from the router

##### Importance
```
yes
```

**Comment by thomas-mangin** [comment]
I believe this was fixed on master. It would be nice if you could check.

**Comment by thomas-mangin** [comment]
For clarity, a similar issue was raised and the line with the problem re-written and I believe the fix is from 4.1.0 with the changelog message
```
 * Fix: do not crash when trying to be helpful in presenting notification message
   reported by: Adam Jacob Muller
```

However from the issue https://github.com/Exa-Networks/exabgp/issues/878, the fix was not tested.

**Comment by pettai** [comment]
Aha, I looked at the release notes for IPv6-related issues and didn't map it to that one.
Then I will try to test it later then I have the time to rebuild ExaBGP package.
Thx for now

**Comment by pettai** [comment]
The good news is that the crash seems fixed as you suggested. 
But the bad news is IPv6 peering isn't established correctly after again.
It lies around in Active to Connect state, and finally get Establ just to drop back to Active state again.

``` exabgp[18642]: 21:20:39 | 18642  | reactor         | connected to peer-2 with outgoing-22 2001:6b0::xx-2001:6b0::yy ```

On the remote side:

```
{master}
pettai@tug-r3-re0> show bgp summary | grep <localASN>
2001:6b0::yy       ASN          2          4       0      43          32 Active


pettai@juniper-re0> show bgp summary | grep <localASN>
2001:6b0::yy       ASN          0          0       0      43          56 Connect

{master}
pettai@juniper-re0> show bgp summary | grep <localASN>
2001:6b0::yy       ASN          1          3       0      43           0 Establ

{master}
pettai@juniper-re0> show bgp summary | grep <localASN>
2001:6b0::yy       ASN          2          4       0      44           2 Active
```

**Comment by thomas-mangin** [comment]
Could I please have see the content of exabgp logs when run with -d ?

**Comment by pettai** [comment]
Hi again,
The output is quite verbose... can I send it to you via email or other ways?

**Comment by thomas-mangin** [comment]
Sure, email is good

**Comment by pettai** [comment]
(I posted a question about your email address in the contact form on your webpage, because I can't find it)

**Comment by pettai** [comment]
Email sent now

**Comment by thomas-mangin** [comment]
ack - will look into it asap.. seems to be caused by:
```
19:11:57 | 15962  | outgoing-9    | peer reset, message [notification
received (3,5)] error['ascii' codec can't decode byte 0xb0 in position 6
: ordinal not in range(128)]
```

**Comment by thomas-mangin** [comment]
There is more than one issue. I will need to look into this a bit later
```
19:11:42 | 15962  | outgoing-1    | << message of type NOTIFICATION
19:11:42 | 15962  | outgoing-1    | peer reset, message [notification
received (3,3)] error[UPDATE message error / Missing Well-known Attribut
e / ]
```

**Comment by pettai** [comment]
ok thanks

**Comment by thomas-mangin** [comment]
A number of comments:
 - some IPv4 routes are announced to peer for which IPv4 unicast was not setup part of the Multi-Protocol negotiation (if you are going to announce IPv4 routes please be explicit about it on both side of the router)
 - I am not sure without spending time in the code that "next-hop self" works for IPv6 .. it just may
 - The session is dropping as one of the peer is NOT configured for IPv6 but still IPv6 routes are sent to it - the BGP session dropping is to be expected in this case, this is the RFC defined behaviour
 - The next session drop comes from a notification from the peer that the route sent is invalid (ExaBGP let you generate what you want) Missing Well-known Attribute. I do not have the code you use to generate the route so I can not tell you what is missing (most likely AS-PATH as it is an ebgp session) 
 - The next notification seems to come as we are retrying to connect fast and the peer has not yet cleaned up the previous session (it works for the next try)

Any further issue after that could be many things not related to ExaBGP .. If we can first fix these, I will happily review the next things then.

**Comment by pettai** [comment]
After specifying type at the Exa side, it no longer seems to be a problem, so I think we can close this issue. The original issue that caused a crash was resolved by 4.1.10.
Thx
